File: templates/python/plugins/langgraph_plugin.py
"""
LangGraph → Borgkit Plugin
──────────────────────────────────────────────────────────────────────────────
Wraps a compiled LangGraph graph (or any callable with a `.invoke()` method)
so it appears as a standard Borgkit IAgent on the mesh.

Capability extraction strategy
──────────────────────────────
LangGraph has no single universal tool registry, so we use four strategies
in priority order:

  1. Explicit map in PluginConfig.capability_map
       { "borgkit_cap_name": "node_or_tool_name" }
  2. Tools bound on the graph's LLM node  (most common ReAct / Tool-Use pattern)
       graph.nodes["agent"].bound.tools  or  graph.nodes["agent"].tools
  3. tools= kwarg passed directly to LangGraphPlugin
  4. Single-capability fallback: one capability = one graph invocation

Usage
─────
  from plugins.langgraph_plugin import LangGraphPlugin, LangGraphPluginConfig
  from plugins.base import PluginConfig

  config = LangGraphPluginConfig(
      agent_id   = "borgkit://agent/weather",
      name       = "WeatherAgent",
      version    = "1.0.0",
      tags       = ["weather", "langraph"],
      # optional: expose each tool as a separate Borgkit capability
      expose_tools_as_capabilities = True,
  )
  plugin = LangGraphPlugin(config)
  agent  = plugin.wrap(compiled_graph)

  await agent.register_discovery()

  # Borgkit will now call agent.handle_request(req)
  # e.g. req.capability == "getWeather", req.payload == {"city": "London"}

Install deps:
  pip install langgraph langchain-core
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Optional

from interfaces import AgentRequest, AgentResponse
from plugins.base import BorgkitPlugin, CapabilityDescriptor, PluginConfig


# ── extended config ───────────────────────────────────────────────────────────

@dataclass
class LangGraphPluginConfig(PluginConfig):
    """LangGraph-specific configuration (extends PluginConfig)."""

    # If True, each tool in the graph is exposed as a separate Borgkit capability.
    # If False, the entire graph is exposed as one capability named `invoke`.
    expose_tools_as_capabilities: bool = True

    # The LangGraph state key that holds the user input (default: "messages")
    input_key: str = "messages"

    # The LangGraph state key where we read the final output (default: "messages")
    output_key: str = "messages"

    # Node name to inspect for bound tools (default: "agent" or "tools")
    agent_node_name: str = "agent"

    # Stream output instead of waiting for full result
    stream: bool = False

    # Recursion limit passed to graph.invoke()
    recursion_limit: int = 25


# ── plugin implementation ─────────────────────────────────────────────────────

class LangGraphPlugin(BorgkitPlugin):
    """
    Borgkit plugin for LangGraph agents.

    Accepts any LangGraph `CompiledGraph` (or duck-typed equivalent).
    """

    def __init__(
        self,
        config: LangGraphPluginConfig,
        tools: Optional[list] = None,
    ):
        super().__init__(config)
        self._lg_config    = config
        self._explicit_tools = tools or []

    # ── capability extraction ─────────────────────────────────────────────────

    def extract_capabilities(self, agent: Any) -> list[CapabilityDescriptor]:
        """
        Walk the graph looking for tools; fall back to a single `invoke` cap.
        """
        if self._lg_config.expose_tools_as_capabilities:
            tools = self._discover_tools(agent)
            if tools:
                return [self._tool_to_descriptor(t) for t in tools]

        # Single-graph capability
        return [CapabilityDescriptor(
            name        = "invoke",
            description = self._lg_config.description or "Invoke the LangGraph agent",
            native_name = "__graph__",
            tags        = self._lg_config.tags,
        )]

    def _discover_tools(self, graph: Any) -> list:
        """
        Try multiple strategies to find tools in a LangGraph graph.
        Returns a list of LangChain BaseTool instances (or duck-types).
        """
        # 1. Explicit tools passed to plugin constructor
        if self._explicit_tools:
            return self._explicit_tools

        # 2. Explicit capability_map → create stub descriptors
        if self.config.capability_map:
            return []  # handled via capability_map in translate_request

        # 3. Inspect graph nodes for bound tools
        try:
            nodes = getattr(graph, 'nodes', {})
            node_name = self._lg_config.agent_node_name
            node = nodes.get(node_name)
            if node:
                # RunnableBinding with tools
                for attr in ('bound', 'runnable'):
                    inner = getattr(node, attr, None)
                    if inner:
                        tools = getattr(inner, 'tools', None)
                        if tools:
                            return list(tools)
                # Direct .tools attribute
                tools = getattr(node, 'tools', None)
                if tools:
                    return list(tools)
        except Exception:
            pass

        # 4. graph.tools attribute (some custom patterns)
        try:
            tools = getattr(graph, 'tools', None)
            if tools:
                return list(tools)
        except Exception:
            pass

        return []

    @staticmethod
    def _tool_to_descriptor(tool: Any) -> CapabilityDescriptor:
        """Convert a LangChain BaseTool to a CapabilityDescriptor."""
        name = getattr(tool, 'name', str(tool))
        desc = getattr(tool, 'description', '')

        # Try to extract JSON schema from tool args_schema (Pydantic model)
        input_schema = None
        try:
            schema_model = getattr(tool, 'args_schema', None)
            if schema_model and hasattr(schema_model, 'schema'):
                input_schema = schema_model.schema()
            elif schema_model and hasattr(schema_model, 'model_json_schema'):
                input_schema = schema_model.model_json_schema()
        except Exception:
            pass

        return CapabilityDescriptor(
            name         = name,
            description  = desc,
            native_name  = name,
            input_schema = input_schema,
            tags         = [],
        )

    # ── request translation ───────────────────────────────────────────────────

    def translate_request(
        self,
        req: AgentRequest,
        descriptor: CapabilityDescriptor,
    ) -> dict:
        """
        Build a LangGraph-compatible state dict from an AgentRequest.

        Strategy:
          - Single graph invocation: wrap payload as a HumanMessage
          - Per-tool invocation: wrap as a ToolCall message
        """
        # Resolve native tool name via capability_map or descriptor
        native = (
            self.config.capability_map.get(req.capability)
            or descriptor.native_name
            or req.capability
        )

        input_key = self._lg_config.input_key

        if native == "__graph__":
            # Whole-graph invocation: treat payload as a human message
            from langchain_core.messages import HumanMessage
            content = (
                req.payload.get("message")
                or req.payload.get("input")
                or json.dumps(req.payload)
            )
            return {
                input_key: [HumanMessage(content=content)],
                "__borgkit_request_id__": req.request_id,
                "__borgkit_capability__": req.capability,
                "__borgkit_from__": req.from_id,
            }
        else:
            # Tool-specific invocation: inject a tool call into the messages
            from langchain_core.messages import HumanMessage
            tool_input = json.dumps(req.payload)
            return {
                input_key: [HumanMessage(content=tool_input)],
                "__borgkit_request_id__": req.request_id,
                "__borgkit_tool__": native,
            }

    # ── response translation ──────────────────────────────────────────────────

    def translate_response(self, native_result: Any, request_id: str) -> AgentResponse:
        """
        Convert LangGraph's state dict output into a Borgkit AgentResponse.
        Handles both dict states (standard) and BaseMessage list outputs.
        """
        try:
            output_key = self._lg_config.output_key

            # State dict
            if isinstance(native_result, dict):
                messages = native_result.get(output_key, [])
                content  = self._extract_content(messages)
                raw      = native_result
            # Direct message list
            elif isinstance(native_result, list):
                content = self._extract_content(native_result)
                raw     = {"messages": native_result}
            else:
                content = str(native_result)
                raw     = {"raw": str(native_result)}

            return AgentResponse.success(request_id, {
                "content": content,
                "raw":     raw,
            })
        except Exception as e:
            return AgentResponse.error(request_id, f"Response translation failed: {e}")

    @staticmethod
    def _extract_content(messages: list) -> str:
        """Pull text from the last non-tool AI message."""
        if not messages:
            return ""
        # Walk backwards to find the last AIMessage
        for msg in reversed(messages):
            kind = type(msg).__name__
            if kind in ("AIMessage", "AIMessageChunk"):
                content = getattr(msg, 'content', '')
                if isinstance(content, list):   # multimodal
                    return " ".join(
                        c.get("text", "") if isinstance(c, dict) else str(c)
                        for c in content
                    )
                return str(content)
        # Fallback: last message content
        last = messages[-1]
        return str(getattr(last, 'content', last))

    # ── native invocation ─────────────────────────────────────────────────────

    async def invoke_native(
        self,
        agent: Any,
        descriptor: CapabilityDescriptor,
        native_input: dict,
    ) -> Any:
        """
        Invoke the LangGraph compiled graph.
        Supports both sync (.invoke) and async (.ainvoke) graphs.
        """
        config = {"recursion_limit": self._lg_config.recursion_limit}

        if self._lg_config.stream:
            return await self._stream_invoke(agent, native_input, config)

        # Prefer async
        if hasattr(agent, 'ainvoke'):
            return await agent.ainvoke(native_input, config=config)

        # Sync fallback (runs in threadpool to avoid blocking event loop)
        import asyncio
        return await asyncio.get_event_loop().run_in_executor(
            None, lambda: agent.invoke(native_input, config=config)
        )

    async def _stream_invoke(self, agent: Any, inp: dict, config: dict) -> Any:
        """Collect all streamed chunks and merge them."""
        result = {}
        if hasattr(agent, 'astream'):
            async for chunk in agent.astream(inp, config=config):
                if isinstance(chunk, dict):
                    result.update(chunk)
        return result or inp


# ── convenience function ──────────────────────────────────────────────────────

def wrap_langgraph(
    graph:   Any,
    name:    str,
    agent_id: str,
    owner:   str  = "0xYourWalletAddress",
    version: str  = "0.1.0",
    tags:    list[str] | None = None,
    tools:   list | None = None,
    expose_tools: bool = True,
    **kwargs,
) -> "WrappedAgent":
    """
    One-liner helper to wrap a LangGraph graph.

    Example:
      from plugins.langgraph_plugin import wrap_langgraph
      agent = wrap_langgraph(
          graph    = compiled_graph,
          name     = "MyResearcher",
          agent_id = "borgkit://agent/researcher",
          tags     = ["research", "web"],
      )
      await agent.register_discovery()
    """
    from plugins.base import WrappedAgent  # re-export type hint
    config = LangGraphPluginConfig(
        agent_id=agent_id, name=name, owner=owner,
        version=version, tags=tags or [],
        expose_tools_as_capabilities=expose_tools,
        **kwargs,
    )
    return LangGraphPlugin(config, tools=tools or []).wrap(graph)

File: templates/typescript/plugins/LangGraphPlugin.ts
/**
 * LangGraph.js → Borgkit Plugin (TypeScript)
 * ─────────────────────────────────────────────────────────────────────────────
 * Wraps a compiled LangGraph `CompiledGraph` (or any object with `.invoke()`)
 * so it appears as a standard Borgkit IAgent on the mesh.
 *
 * Capability extraction strategy (priority order):
 *   1. Explicit `capabilityMap` in config
 *   2. Tools bound on the graph's agent node  (.nodes["agent"].bound.tools)
 *   3. `tools` passed directly to LangGraphPlugin constructor
 *   4. Single `invoke` capability (whole-graph fallback)
 *
 * Usage:
 *   import { LangGraphPlugin } from './plugins/LangGraphPlugin';
 *
 *   const plugin = new LangGraphPlugin({
 *     agentId: 'borgkit://agent/researcher',
 *     name:    'ResearchAgent',
 *     version: '1.0.0',
 *     tags:    ['research', 'web'],
 *     exposeToolsAsCapabilities: true,
 *   });
 *
 *   const agent = plugin.wrap(compiledGraph);
 *   await agent.registerDiscovery();
 *
 * Install: npm install @langchain/langgraph @langchain/core
 */

import { AgentRequest }         from '../interfaces/IAgentRequest';
import { AgentResponse }        from '../interfaces/IAgentResponse';
import { BorgkitPlugin, PluginConfig, CapabilityDescriptor } from './IPlugin';

// ── extended config ───────────────────────────────────────────────────────────

export interface LangGraphPluginConfig extends PluginConfig {
  /** Expose each tool as a separate capability (default: true) */
  exposeToolsAsCapabilities?: boolean;
  /** LangGraph state key for user input messages (default: 'messages') */
  inputKey?: string;
  /** LangGraph state key for output messages (default: 'messages') */
  outputKey?: string;
  /** Graph node name to inspect for bound tools (default: 'agent') */
  agentNodeName?: string;
  /** LangGraph recursion limit (default: 25) */
  recursionLimit?: number;
  /** Stream output instead of single invoke (default: false) */
  stream?: boolean;
}

// ── native input / output types ───────────────────────────────────────────────

interface LangGraphInput {
  [key: string]: unknown;
  __borgkitRequestId__: string;
  __borgkitCapability__: string;
}

interface LangGraphOutput {
  [key: string]: unknown;
}

// ── plugin ────────────────────────────────────────────────────────────────────

export class LangGraphPlugin extends BorgkitPlugin<unknown, LangGraphInput, LangGraphOutput> {
  private readonly lgConfig: Required<LangGraphPluginConfig>;
  private readonly explicitTools: unknown[];

  constructor(config: LangGraphPluginConfig, tools: unknown[] = []) {
    super(config);
    this.lgConfig = {
      exposeToolsAsCapabilities: true,
      inputKey:       'messages',
      outputKey:      'messages',
      agentNodeName:  'agent',
      recursionLimit: 25,
      stream:         false,
      host:           'localhost',
      port:           6174,
      protocol:       'http',
      tls:            false,
      discoveryType:  'local',
      tags:           [],
      capabilityMap:  {},
      ...config,
    } as Required<LangGraphPluginConfig>;
    this.explicitTools = tools;
  }

  // ── capability extraction ──────────────────────────────────────────────────

  extractCapabilities(graph: unknown): CapabilityDescriptor[] {
    if (this.lgConfig.exposeToolsAsCapabilities) {
      const tools = this.discoverTools(graph);
      if (tools.length > 0) {
        return tools.map(t => this.toolToDescriptor(t));
      }
    }

    // Single graph invocation fallback
    return [{
      name:        'invoke',
      description: this.config.description ?? 'Invoke the LangGraph agent',
      nativeName:  '__graph__',
      tags:        this.config.tags ?? [],
    }];
  }

  private discoverTools(graph: unknown): unknown[] {
    if (this.explicitTools.length > 0) return this.explicitTools;

    const g = graph as Record<string, unknown>;

    // Check graph.nodes["agent"].bound.tools or .tools
    try {
      const nodes = g['nodes'] as Record<string, unknown> | undefined;
      if (nodes) {
        const node = nodes[this.lgConfig.agentNodeName] as Record<string, unknown> | undefined;
        if (node) {
          for (const attr of ['bound', 'runnable']) {
            const inner = node[attr] as Record<string, unknown> | undefined;
            if (inner?.['tools']) return inner['tools'] as unknown[];
          }
          if (node['tools']) return node['tools'] as unknown[];
        }
      }
    } catch { /* ignore */ }

    // Direct graph.tools
    if (g['tools']) return g['tools'] as unknown[];

    return [];
  }

  private toolToDescriptor(tool: unknown): CapabilityDescriptor {
    const t = tool as Record<string, unknown>;
    const name = (t['name'] as string) ?? String(tool);
    const desc = (t['description'] as string) ?? '';

    let inputSchema: Record<string, unknown> | undefined;
    try {
      const schemaModel = t['schema'] as Record<string, unknown> | undefined;
      if (schemaModel) inputSchema = schemaModel as Record<string, unknown>;
    } catch { /* ignore */ }

    return { name, description: desc, nativeName: name, inputSchema };
  }

  // ── request translation ────────────────────────────────────────────────────

  translateRequest(req: AgentRequest, descriptor: CapabilityDescriptor): LangGraphInput {
    const native = this.config.capabilityMap?.[req.capability] ?? descriptor.nativeName;

    if (native === '__graph__') {
      const content = (req.payload['message'] ?? req.payload['input'] ?? JSON.stringify(req.payload)) as string;
      return {
        [this.lgConfig.inputKey]: [{ role: 'human', content }],
        __borgkitRequestId__:    req.requestId,
        __borgkitCapability__:   req.capability,
      };
    }

    // Tool-specific: craft a prompt containing the tool call args
    const argsStr = JSON.stringify(req.payload, null, 2);
    return {
      [this.lgConfig.inputKey]: [{
        role:    'human',
        content: `Call tool \`${native}\` with:\n${argsStr}`,
      }],
      __borgkitRequestId__:  req.requestId,
      __borgkitCapability__: req.capability,
      __borgkitTool__:       native,
    };
  }

  // ── response translation ───────────────────────────────────────────────────

  translateResponse(result: LangGraphOutput, requestId: string): AgentResponse {
    try {
      const messages = result[this.lgConfig.outputKey] as unknown[];
      const content  = messages ? this.extractContent(messages) : String(result);
      return { requestId, status: 'success', result: { content, raw: result }, timestamp: Date.now() };
    } catch (e) {
      return { requestId, status: 'error', errorMessage: `Response translation failed: ${e}` };
    }
  }

  private extractContent(messages: unknown[]): string {
    // Walk backwards to find the last AI message
    for (let i = messages.length - 1; i >= 0; i--) {
      const msg = messages[i] as Record<string, unknown>;
      const type = msg['type'] as string | undefined;
      if (type === 'ai' || type === 'AIMessage' || msg['role'] === 'assistant') {
        const content = msg['content'];
        if (typeof content === 'string') return content;
        if (Array.isArray(content)) {
          return content.map(c => (c as Record<string, unknown>)['text'] ?? String(c)).join(' ');
        }
      }
    }
    const last = messages[messages.length - 1] as Record<string, unknown> | undefined;
    return last ? String(last['content'] ?? last) : '';
  }

  // ── native invocation ──────────────────────────────────────────────────────

  async invokeNative(
    graph:      unknown,
    _d:         CapabilityDescriptor,
    input:      LangGraphInput,
  ): Promise<LangGraphOutput> {
    const g = graph as Record<string, unknown>;
    const cfg = { recursionLimit: this.lgConfig.recursionLimit };

    if (this.lgConfig.stream && typeof g['stream'] === 'function') {
      const chunks: LangGraphOutput = {};
      for await (const chunk of (g['stream'] as Function)(input, cfg)) {
        if (typeof chunk === 'object') Object.assign(chunks, chunk);
      }
      return chunks;
    }

    if (typeof g['invoke'] === 'function') {
      return (g['invoke'] as Function)(input, cfg);
    }
    if (typeof g['ainvoke'] === 'function') {
      return (g['ainvoke'] as Function)(input, cfg);
    }

    throw new Error('Graph does not have invoke() or ainvoke() method');
  }
}

// ── one-liner helper ──────────────────────────────────────────────────────────

export function wrapLangGraph(
  graph:    unknown,
  config:   LangGraphPluginConfig,
  tools?:   unknown[],
) {
  return new LangGraphPlugin(config, tools).wrap(graph);
}
